data/trie/trieStorageManager.go

```diff
@@ -0,0 +1,329 @@
+	var keys [][]byte
+	isSnapshotsBufferEmpty := false
+	for !isSnapshotsBufferEmpty {
+		tsm.storageOperationMutex.Lock()
+
+		snapshot := tsm.snapshotsBuffer.getFirst()
+		tr, err := newSnapshotTrie(tsm.db, msh, hsh, snapshot.rootHash)
+		if err != nil {
+			log.Error("trie storage manager: newSnapshotTrie", "error", err.Error())
+			return
+		}
+		db := tsm.getSnapshotDb(snapshot.newDb)
+
+		tsm.storageOperationMutex.Unlock()
+
+		err = tr.root.commit(true, 0, tsm.db, db)
+		if err != nil {
+			log.Error("trie storage manager: commit", "error", err.Error())
+			return
+		}
+
+		tsm.storageOperationMutex.Lock()
+		tsm.snapshotsBuffer.removeFirst()
+		isSnapshotsBufferEmpty = tsm.snapshotsBuffer.len() == 0
+		if isSnapshotsBufferEmpty {
+			keys = tsm.pruningBuffer
+			tsm.pruningBuffer = make([][]byte, 0)
+		}
+		tsm.storageOperationMutex.Unlock()
+	}
+
+	tsm.removeKeysFromDb(keys)
+}
+
+func (tsm *trieStorageManager) removeKeysFromDb(keys [][]byte) {
+	for i := range keys {
+		tsm.storageOperationMutex.Lock()
+		err := tsm.removeFromDb(keys[i])
+		if err != nil {
+			log.Error("trie storage manager: removeKeysFromDb", "error", err.Error())
+		}
+		tsm.storageOperationMutex.Unlock()
+	}
+}
+
+func (tsm *trieStorageManager) getSnapshotDb(newDb bool) data.DBWriteCacher {
+	createNewDb := newDb || len(tsm.snapshots) == 0
+	if !createNewDb {
+		return tsm.snapshots[len(tsm.snapshots)-1]
+	}
+
+	db, err := tsm.newSnapshotDb()
+	if err != nil {
+		log.Error("trie storage manager: getSnapshotDb", "error", err.Error())
+		return nil
+	}
+
+	if len(tsm.snapshots) > maxSnapshots {
+		tsm.removeSnapshot()
+	}
+
+	return db
+}
+
+func (tsm *trieStorageManager) removeSnapshot() {
+	dbUniqueId := strconv.Itoa(tsm.snapshotId - len(tsm.snapshots))
+
+	err := tsm.snapshots[0].Close()
+	if err != nil {
+		log.Error("trie storage manager: removeSnapshot", "error", err.Error())
+		return
+	}
+	tsm.snapshots = tsm.snapshots[1:]
+
+	removePath := path.Join(tsm.snapshotDbCfg.FilePath, dbUniqueId)
+	go removeDirectory(removePath)
+}
+
+func removeDirectory(path string) {
+	err := os.RemoveAll(path)
+	if err != nil {
+		log.Error(err.Error())
+	}
+}
+
+func newSnapshotTrie(
+	db data.DBWriteCacher,
+	msh marshal.Marshalizer,
+	hsh hashing.Hasher,
+	rootHash []byte,
+) (*patriciaMerkleTrie, error) {
+	newRoot, err := getNodeFromDBAndDecode(rootHash, db, msh, hsh)
+	if err != nil {
+		return nil, err
+	}
+
+	trieStorage, err := NewTrieStorageManager(db, &config.DBConfig{}, &mock.EvictionWaitingList{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &patriciaMerkleTrie{
+		root:        newRoot,
+		trieStorage: trieStorage,
+		marshalizer: msh,
+		hasher:      hsh,
+	}, nil
+}
+
+func (tsm *trieStorageManager) newSnapshotDb() (storage.Persister, error) {
+	snapshotPath := path.Join(tsm.snapshotDbCfg.FilePath, strconv.Itoa(tsm.snapshotId))
+	_, err := os.Stat(snapshotPath)
+	for !os.IsNotExist(err) {
```

Comment:
Please extract into a function or variable (`!os.IsNotExist(err)`)- it's hard to make sense of every aspect of what is treated here.

Reply:
Done.